File: server/internal/risk/presidiofp/classify.go
// Package presidiofp classifies Presidio PII findings as false positives. It
// holds the per-entity false-positive catalogs (reserved/placeholder IPs and
// emails, cloud/CDN ASN attribution, NHS number validity, retired recognizers)
// and the dispatch over them. It is a leaf domain package with no Temporal or
// activity dependencies, so it can be reused from the scanner, from offline
// tools, or anywhere a stored finding needs to be re-evaluated.
package presidiofp

import "strings"

// EntityType is Presidio's UPPER_SNAKE entity name (e.g. "IP_ADDRESS").
type EntityType = string

const (
	EntityTypeEmailAddress    EntityType = "EMAIL_ADDRESS"
	EntityTypeIPAddress       EntityType = "IP_ADDRESS"
	EntityTypeUKNHS           EntityType = "UK_NHS"
	EntityTypeUSDriverLicense EntityType = "US_DRIVER_LICENSE"
)

// rulePrefix is the canonical rule_id prefix for Presidio PII findings. It
// mirrors risk_analysis.CanonicalPresidioRuleID's grammar ("pii." + lowercased
// entity); keep the two in sync.
const rulePrefix = "pii."

// Reason returns the catalog reason a Presidio match of the given entity type
// is treated as noise, or "" when it is a real finding. Entity types without a
// catalog always return "".
//
// This is the value-only view: it judges the matched text on its own. Callers
// that hold the text the match was found in should prefer ReasonInContext,
// which additionally applies the catalogs that need surrounding text.
func Reason(entityType, match string) string {
	return ReasonInContext(entityType, match, "")
}

// ReasonInContext is Reason with the payload the match was found in. Passing an
// empty text is equivalent to calling Reason: no finding is ever suppressed for
// missing context, only for context that is present and carries no signal.
func ReasonInContext(entityType, match, text string) string {
	if reason := retiredReason(entityType); reason != "" {
		return reason
	}
	switch entityType {
	case EntityTypeIPAddress:
		return nonPIIIPReason(strings.TrimSpace(match))
	case EntityTypeEmailAddress:
		return nonPIIEmailReason(match)
	case EntityTypeUKNHS:
		if reason := nonNHSReason(match); reason != "" {
			return reason
		}
		return nhsContextReason(text)
	default:
		return ""
	}
}

// ReasonByRuleID is Reason keyed by a stored finding's canonical rule_id
// (e.g. "pii.ip_address"), for re-evaluating persisted findings. Rule ids
// outside the catalogs always return "".
func ReasonByRuleID(ruleID, match string) string {
	return Reason(entityTypeForRuleID(ruleID), match)
}

// ReasonByRuleIDInContext is ReasonInContext keyed by a stored finding's
// canonical rule_id, for the offline sweep that re-reads the message a finding
// was anchored to.
func ReasonByRuleIDInContext(ruleID, match, text string) string {
	return ReasonInContext(entityTypeForRuleID(ruleID), match, text)
}

// RuleIDs returns the canonical rule ids whose entity types have a catalog.
// Callers re-scanning stored findings can use it to read only rows that could
// possibly be reclassified. Keep in sync with the switch in ReasonInContext.
func RuleIDs() []string {
	ids := []string{
		ruleIDForEntity(EntityTypeIPAddress),
		ruleIDForEntity(EntityTypeEmailAddress),
		ruleIDForEntity(EntityTypeUKNHS),
	}
	for _, entity := range RetiredEntityTypes() {
		ids = append(ids, ruleIDForEntity(entity))
	}
	return ids
}

// ContextRuleIDs returns the subset of RuleIDs whose classification can change
// once the surrounding text is supplied. The sweep uses it to decide which rows
// are worth re-reading a message for.
func ContextRuleIDs() []string {
	return []string{ruleIDForEntity(EntityTypeUKNHS)}
}

// ruleIDForEntity maps a Presidio entity type to its canonical rule_id.
func ruleIDForEntity(entity EntityType) string {
	return rulePrefix + strings.ToLower(entity)
}

// entityTypeForRuleID inverts ruleIDForEntity: "pii.ip_address" -> "IP_ADDRESS".
// Returns "" for non-PII rule ids.
func entityTypeForRuleID(ruleID string) string {
	rest, ok := strings.CutPrefix(ruleID, rulePrefix)
	if !ok {
		return ""
	}
	return strings.ToUpper(rest)
}

File: pystreams/src/pystreams/risk/presidiofp/classify.py
"""Classifies Presidio PII findings as false positives.

Holds the per-entity false-positive catalogs (reserved/placeholder IPs and
emails, cloud/CDN ASN attribution, NHS number validity, retired recognizers) and
the dispatch over them.
"""

from .email import non_pii_email_reason
from .ip import non_pii_ip_reason
from .nhs import nhs_context_reason, non_nhs_reason
from .retired import RETIRED_RECOGNIZERS, retired_reason

# Presidio's UPPER_SNAKE entity names that have a false-positive catalog.
ENTITY_TYPE_EMAIL_ADDRESS = "EMAIL_ADDRESS"
ENTITY_TYPE_IP_ADDRESS = "IP_ADDRESS"
ENTITY_TYPE_UK_NHS = "UK_NHS"

# The canonical rule_id prefix for Presidio PII findings. Mirrors
# ``_canonical_rule_id`` in the handler ("pii." + lowercased entity); keep in sync.
_RULE_PREFIX = "pii."


def reason(entity_type: str, match: str) -> str:
    """Return the catalog reason a Presidio match is treated as noise, or "" when
    it is a real finding. Entity types without a catalog always return "".

    This is the value-only view: it judges the matched text on its own. Callers
    that hold the text the match was found in should prefer ``reason_in_context``,
    which additionally applies the catalogs that need surrounding text.
    """
    return reason_in_context(entity_type, match, "")


def reason_in_context(entity_type: str, match: str, text: str) -> str:
    """``reason`` with the payload the match was found in. Passing an empty
    ``text`` is equivalent to calling ``reason``: no finding is ever suppressed
    for missing context, only for context that is present and carries no signal.
    """
    retired = retired_reason(entity_type)
    if retired:
        return retired
    if entity_type == ENTITY_TYPE_IP_ADDRESS:
        return non_pii_ip_reason(match.strip())
    if entity_type == ENTITY_TYPE_EMAIL_ADDRESS:
        return non_pii_email_reason(match)
    if entity_type == ENTITY_TYPE_UK_NHS:
        return non_nhs_reason(match) or nhs_context_reason(text)
    return ""


def reason_by_rule_id(rule_id: str, match: str) -> str:
    """``reason`` keyed by a stored finding's canonical rule_id (e.g.
    ``pii.ip_address``), for re-evaluating persisted findings. Rule ids outside
    the catalogs always return "".
    """
    return reason(_entity_type_for_rule_id(rule_id), match)


def reason_by_rule_id_in_context(rule_id: str, match: str, text: str) -> str:
    """``reason_in_context`` keyed by a stored finding's canonical rule_id."""
    return reason_in_context(_entity_type_for_rule_id(rule_id), match, text)


def rule_ids() -> list[str]:
    """Return the canonical rule ids whose entity types have a catalog. Keep in
    sync with the dispatch in ``reason_in_context``.
    """
    return [
        _rule_id_for_entity(ENTITY_TYPE_IP_ADDRESS),
        _rule_id_for_entity(ENTITY_TYPE_EMAIL_ADDRESS),
        _rule_id_for_entity(ENTITY_TYPE_UK_NHS),
    ] + [_rule_id_for_entity(entity) for entity in sorted(RETIRED_RECOGNIZERS)]


def context_rule_ids() -> list[str]:
    """Return the subset of ``rule_ids`` whose classification can change once the
    surrounding text is supplied.
    """
    return [_rule_id_for_entity(ENTITY_TYPE_UK_NHS)]


def _rule_id_for_entity(entity: str) -> str:
    return _RULE_PREFIX + entity.lower()


def _entity_type_for_rule_id(rule_id: str) -> str:
    """Invert ``_rule_id_for_entity``: ``pii.ip_address`` -> ``IP_ADDRESS``.
    Returns "" for non-PII rule ids.
    """
    if not rule_id.startswith(_RULE_PREFIX):
        return ""
    return rule_id[len(_RULE_PREFIX) :].upper()
